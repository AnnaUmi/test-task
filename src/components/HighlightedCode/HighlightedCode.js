import React from 'react';
import PropTypes from 'prop-types';
import Prism from 'prismjs';
import 'prismjs/components/prism-jsx';
import 'prismjs/components/prism-markdown';
import { Pre, Code, Token } from './HighlightedCode.styles';

const isToken = t => t instanceof Prism.Token;

export const renderPrismTokens = tokens =>
  tokens.map((token, i) => {
    if (isToken(token)) {
      const { type, content } = token;
      const key = `${type}-${i}`; // no other prop to use as key except index
      return (
        <Token key={key} type={type}>
          {Array.isArray(content) ? renderPrismTokens(content) : content}
        </Token>
      );
    }
    return token;
  });

const HighlightedCode = ({ language, code, isVisible }) => {
  const prismLanguage = Prism.languages[language];
  const prismTokens = Prism.tokenize(code, prismLanguage);

  return (
    isVisible && (
      <Pre>
        <Code>{renderPrismTokens(prismTokens)}</Code>
      </Pre>
    )
  );
};

HighlightedCode.propTypes = {
  language: PropTypes.string,
  code: PropTypes.string.isRequired,
  isVisible: PropTypes.bool,
};

HighlightedCode.defaultProps = {
  language: 'jsx',
  isVisible: false,
};

export default HighlightedCode;
