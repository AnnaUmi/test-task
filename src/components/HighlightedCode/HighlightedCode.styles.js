import styled from 'styled-components';

const codeColors = {
  function: '#FF5555',
  keyword: '#3F7397',
  punctuation: '#535353',
  script: '#3F7397',
  string: '#00263E',
  tag: '#FF5555',
};

export const Pre = styled.pre`
  -webkit-font-smoothing: antialiased;
  -webkit-overflow-scrolling: touch;
  border-radius: 6px;
  font-family: 'Roboto', sans-serif;
  font-size: 15px;
  font-style: normal;
  font-weight: 400;
  background: #f9f9f9;
  border: none;
  box-sizing: border-box;
  color: #00263e;
  display: block;
  height: auto;
  line-height: 1.44;
  margin: 25px 0 0 0;
  overflow: auto;
  padding: 20px;
  white-space: pre;
  width: 100%;
  text-rendering: optimizeLegibility;
`;

export const Code = styled.code`
  font-family: 'Roboto Mono', monospace;
  font-weight: 400;
`;

export const Token = styled.span`
  ${({ type }) => `color: ${codeColors[type]}`};
`;
