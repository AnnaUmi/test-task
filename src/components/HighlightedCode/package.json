{
  "name": "highlighted-code",
  "version": "0.0.0",
  "private": true,
  "main": "./HighlightedCode.js"
}
