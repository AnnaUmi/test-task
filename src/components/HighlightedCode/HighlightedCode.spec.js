import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import HighlightedCode from './HighlightedCode';

describe('HeaderNav', () => {
  let code;
  beforeEach(() => {
    code = `
      <Button variant='standard'>
        My cool button
      </Button>
    `;
  });

  it('should render code component', () => {
    const tree = renderer.create(<HighlightedCode code={code} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
