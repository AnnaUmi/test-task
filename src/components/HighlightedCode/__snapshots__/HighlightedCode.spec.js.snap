// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderNav should render code component 1`] = `null`;
